tmc-langs-framework/src/main/java/fi/helsinki/cs/tmc/langs/LanguagePluginAbstract.java

```diff
@@ -22,17 +34,91 @@ public void prepareSubmission(Path submissionPath, Path destPath) {
 
     @Override
     public void prepareStub(Path path) {
-        throw new UnsupportedOperationException("Not supported yet."); //To change body of generated methods, choose Tools | Templates.
+        exerciseBuilder.prepareStub(path);
     }
 
     @Override
     public void prepareSolution(Path path) {
-        throw new UnsupportedOperationException("Not supported yet."); //To change body of generated methods, choose Tools | Templates.
+        exerciseBuilder.prepareSolution(path);
     }
-
+    
+    /**
+     * Check if the exercise's project type corresponds with the language plugin
+     * type.
+     *
+     * @param path The path to the exercise directory.
+     * @return True if given path is valid directory for this language plugin
+     */
+    protected abstract boolean isExerciseTypeCorrect(Path path);
+    
+    /**
+     *
+     * @param basePath The file path to search in.
+     * @return A list of directories that contain a build file in this language.
+     */
     @Override
-    public boolean isExerciseTypeCorrect(Path path) {
-        throw new UnsupportedOperationException("Not supported yet.");
+    public ImmutableList<Path> findExercises(Path basePath) {
+        File searchPath = new File(basePath.toString());
+        ImmutableList.Builder<Path> listBuilder = new ImmutableList.Builder<>();
+        if (searchPath.exists() && searchPath.isDirectory()) {
+            return search(searchPath, listBuilder);
+        } else {
+            return listBuilder.build();
+        }
+    }
+
+    /**
+     * Search a directory and its subdirectories for build files. If a directory
+     * contains a build file, the directory is added to the list.
+     *
+     * @param file The current file path to search in
+     * @param listBuilder
+     * @return a list of all directories that contain build files for this
+     * language.
+     */
+    private ImmutableList<Path> search(File file, ImmutableList.Builder<Path> listBuilder) {
+        Stack<File> stack = new Stack();
+        // Push the initial directory onto the stack.
+        stack.push(file);
+        // Walk the directories that get added onto the stack.
+        while (!stack.isEmpty()) {
+            File current = stack.pop();
+            if (current.isDirectory()) {
+                // See if current directory contains a build file.
+                if (isExerciseTypeCorrect(current.toPath())) {
+                    listBuilder.add(current.toPath());
+                }
+                for (File temp : current.listFiles()) {
+                    if (temp.isDirectory()) {
+                        stack.push(temp);
+                    }
+                }
+            }
+        }
+        return listBuilder.build();
     }
 
+    /**
+     * Start a process using ProcessBuilder.
+     *
+     * @param args Arguments for starting the process.
+     * @return Possible output of the process.
+     */
+    protected List<String> startProcess(List<String> args) {
```

Comment:
In fact, it'd be nice if we could (without too much work) avoid having `lib` in the test projects, and make their `build.xml` minimal and not have netbeans project files there eiher.